File: apps/api/src/math_agent_api/routers/documents.py
from fastapi import APIRouter, Depends, File, UploadFile
from fastapi.responses import JSONResponse, Response
from sqlalchemy.orm import Session

from math_agent_api.db.session import get_db_session
from math_agent_api.schemas.common import ErrorBody
from math_agent_api.schemas.documents import DocumentSummary, DocumentUploadResponse
from math_agent_api.services.document_service import (
    DocumentValidationError,
    delete_document,
    ingest_pdf_document,
    list_document_summaries,
)

router = APIRouter(prefix="/documents", tags=["documents"])


@router.post("/upload", response_model=DocumentUploadResponse)
async def upload_document(
    file: UploadFile = File(...),
    db: Session = Depends(get_db_session),
) -> DocumentUploadResponse | JSONResponse:
    try:
        content = await file.read()
        document = ingest_pdf_document(
            db=db,
            content=content,
            filename=file.filename or "uploaded.pdf",
            content_type=file.content_type,
        )
    except DocumentValidationError as exc:
        return JSONResponse(
            status_code=400,
            content={
                "error": ErrorBody(
                    code="document_validation_error",
                    message=str(exc),
                ).model_dump(mode="json")
            },
        )
    return DocumentUploadResponse(document=document)


@router.get("", response_model=list[DocumentSummary])
async def list_documents(db: Session = Depends(get_db_session)) -> list[DocumentSummary]:
    return list_document_summaries(db)


@router.delete("/{document_id}", response_model=None)
async def remove_document(document_id: str, db: Session = Depends(get_db_session)) -> Response | JSONResponse:
    deleted = delete_document(db, document_id)
    if not deleted:
        return JSONResponse(
            status_code=404,
            content={
                "error": ErrorBody(
                    code="document_not_found",
                    message="Document not found.",
                ).model_dump(mode="json")
            },
        )
    return Response(status_code=204)

File: apps/web/src/lib/api/documents.ts
import { API_BASE_URL } from "./config";
import type { DocumentSummary, DocumentUploadResponse } from "@/types/chat";

export async function uploadDocument(file: File): Promise<DocumentSummary> {
  const form = new FormData();
  form.append("file", file);

  const response = await fetch(`${API_BASE_URL}/documents/upload`, {
    method: "POST",
    body: form
  });

  const payload = (await response.json()) as unknown;
  if (!response.ok) {
    throw new Error(extractErrorMessage(payload, `Document upload failed with ${response.status}`));
  }

  return (payload as DocumentUploadResponse).document;
}

export async function listDocuments(): Promise<DocumentSummary[]> {
  const response = await fetch(`${API_BASE_URL}/documents`);
  if (!response.ok) {
    throw new Error(`Document list failed with ${response.status}`);
  }
  return (await response.json()) as DocumentSummary[];
}

export async function deleteDocument(documentId: string): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/documents/${encodeURIComponent(documentId)}`, {
    method: "DELETE"
  });
  if (!response.ok) {
    const payload = (await response.json().catch(() => null)) as unknown;
    throw new Error(extractErrorMessage(payload, `Document delete failed with ${response.status}`));
  }
}

function extractErrorMessage(payload: unknown, fallback: string): string {
  if (
    typeof payload === "object" &&
    payload !== null &&
    "error" in payload &&
    typeof payload.error === "object" &&
    payload.error !== null &&
    "message" in payload.error &&
    typeof payload.error.message === "string"
  ) {
    return payload.error.message;
  }

  return fallback;
}
